source/extensions/common/wasm/wasm.cc

```diff
@@ -80,6 +80,47 @@ class SharedData {
     return true;
   }
 
+  void registerQueue(absl::string_view queue_name, absl::string_view vm_id, uint32_t context_id,
+                     Event::Dispatcher& dispatcher) {
+    absl::WriterMutexLock l(&mutex);
+    auto& q = queues[queue_name];
+    q.vm_id = std::string(vm_id);
+    q.context_id = context_id;
+    q.dispatcher = &dispatcher;
+    // Preserve any existing data.
+  }
+
+  // Returns: data, eos.
+  std::pair<std::string, bool> dequeue(absl::string_view queue_name) {
+    absl::ReaderMutexLock l(&mutex);
+    auto it = queues.find(queue_name);
+    if (it == queues.end() || it->second.queue.empty()) {
+      return {"", true};
+    }
+    auto data = it->second.queue.front();
+    it->second.queue.pop_front();
+    return {std::move(data), it->second.queue.empty()};
+  }
+
+  bool enqueue(absl::string_view queue_name, absl::string_view value) {
+    absl::WriterMutexLock l(&mutex);
+    auto it = queues.find(queue_name);
+    if (it == queues.end()) {
+      return false;
+    }
+    it->second.queue.push_back(std::string(value));
+    auto vm_id = it->second.vm_id;
+    auto context_id = it->second.context_id;
+    auto qn = std::string(queue_name);
+    it->second.dispatcher->post([vm_id, context_id, qn] {
+      auto wasm = getThreadLocalWasmOrNull(vm_id);
```

Comment:
This means vm_id cannot be empty when using queue? There will only be one consumer on this queue right? Can't we just keep a pointer to context of that consumer VM?

Reply:
The context can disappear at any time, so we have to look it up.  The empty string is acceptable for the vm_id if you don't want to define one.